compiler/modules/front/codeGen.bal

```diff
@@ -769,40 +804,106 @@ function codeGenIfElseStmt(CodeGenContext cx, bir:BasicBlock startBlock, Environ
-    bir:Register narrowed = cx.createVarRegister(narrowedType, binding.name, pos);
-    bir:CondNarrowInsn insn = {
-        result: narrowed,
-        operand: binding.reg,
-        basis,
-        pos
-    };
-    bb.insns.push(insn);
-    Binding bindings = {
-        name: binding.name,
-        reg: narrowed,
-        isFinal: binding.isFinal,
-        prev: env.bindings,
-        unnarrowed: unnarrowBinding(binding)
-    };
-    return { bindings, assignments: env.assignments };
+// Union narrowings exist in both lists, drop others
+function unionStmtNarrowings(StmtNarrowing[] snl1, StmtNarrowing[] snl2) returns StmtNarrowing[] {
+    StmtNarrowing[] result = [];
+    foreach var sn1 in snl1 {
+        foreach var sn2 in snl2 {
+            if sn1.binding.name == sn2.binding.name {
+                result.push({
+                    basis: { or: [sn1.basis, sn2.basis] },
+                    binding: sn1.binding,
+                    ifCompletesNormally: t:union(sn1.ifCompletesNormally, sn2.ifCompletesNormally)
+                });
+                break;
+            }
+        }
+    }
+    return result;
+}
+
+// Modifies dest. If already exists in dest, intersect, else append.
+function addIntersectStmtNarrowings(StmtNarrowing[] dest, StmtNarrowing[] src) {
+    foreach var s in src {
+        boolean added = false;
+        foreach int i in 0 ..< dest.length() {
+            StmtNarrowing d = dest[i];
+            if s.binding.name == d.binding.name {
+                dest[i] = {
+                    basis: { and: [s.basis, d.basis] },
+                    binding: s.binding,
+                    // Due to folding, lexically successive one (lets say `s`) is always a subtype. So this is same as s.ifCompletesNormally
+                    ifCompletesNormally: t:intersect(s.ifCompletesNormally, d.ifCompletesNormally)
+                };
+                added = true;
+                break;
+            }
+        }
+        if !added {
+            dest.push(s);
+        }
+    }
+}
+
+function optStmtNarrowingsFromExprNarrowing(ExprNarrowing? narrowing, boolean condition) returns StmtNarrowing? {
+    if narrowing == () {
+        return ();
+    }
+    else {
+        boolean insnResult = condition == !narrowing.negated;
+        // JBUG #33303 without parentheses this gets a parse error
+        t:SemType narrowedType = insnResult ? (narrowing.ifTrue) : narrowing.ifFalse;
+        if narrowedType === t:NEVER {
+            panic err:impossible("narrowed to never type");
+        }
+        bir:Result basis = { insn: narrowing.testInsn, result: insnResult };
+        return { basis, binding: narrowing.binding, ifCompletesNormally: narrowedType };
+    }
+}
+
+function codeGenNarrowings(CodeGenContext cx, bir:BasicBlock bb, Environment env, StmtNarrowing[] narrowings, Position pos) {
```

Comment:
I don't think this should be called `codeGenX`. What it is doing is adding the narrowings to the env. Maybe `addNarrowings`.